src/visibility_tracker.rs

```diff
@@ -606,198 +595,204 @@ fn register_name<'a>(
 /// Collect all public items that are reachable from the crate root and record their parent Ids.
 fn visit_root_reachable_public_items<'a>(
     crate_: &'a Crate,
-    parents: &mut HashMap<&'a Id, HashSet<&'a Id>>,
     traversal_state: &NameResolution<'a>,
-    currently_visited_items: &mut HashSet<&'a Id>,
     item: &'a Item,
-    parent_id: Option<&'a Id>,
-) {
-    match item.visibility {
-        Visibility::Crate | Visibility::Restricted { .. } => {
-            // This item is not public, so we don't need to process it.
-            return;
-        }
-        Visibility::Public => {} // Public item, keep going.
-        Visibility::Default => {
-            // Enum variants, and some impls and methods have default visibility:
-            // they are visible only if the type to which they belong is visible.
-            // However, we don't recurse into non-public items with this function, so
-            // reachable items with default visibility must be public.
-        }
-    }
-
-    let item_parents = parents.entry(&item.id).or_default();
-    if let Some(parent_id) = parent_id {
-        item_parents.insert(parent_id);
+) -> HashMap<&'a Id, Vec<&'a Id>> {
+    #[derive(Debug)]
+    enum Action<'a> {
+        Process {
+            item: &'a Item,
+            parent_id: Option<&'a Id>,
+        },
+        Remove {
+            id: &'a Id,
+        },
     }
 
-    if !currently_visited_items.insert(&item.id) {
-        // We found a cycle in the import graph, and we've already processed this item.
-        // Nothing more to do here.
-        return;
-    }
+    let mut parents: HashMap<&'a Id, Vec<&'a Id>> = Default::default();
+    // Avoid cycles by keeping track of which items we're in the middle of visiting.
+    let mut currently_visited_items: HashSet<&'a str> = Default::default();
+
+    let mut stack = vec![Action::Process {
+        item,
+        parent_id: None,
+    }];
```

Comment:
Similarly here, it would be good to leave an explanatory comment for what's happening: we're simulating recursion with a stack, the different actions mean A and B and correspond to equivalent operations X and Y in the recursive model. The stack-based implementation is faster because of reasons XYZ etc.

All this will be useful context for the next person who might try to update or optimize this code further.

Reply:
Fixed (by removing the refactor)